// eslint-plugin-sonarjs/cjs/S8781/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * You can redistribute and/or modify this program under the terms of
 * the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S8781/javascript
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const generate_meta_js_1 = require("../helpers/generate-meta.js");
const module_js_1 = require("../helpers/module.js");
const mocha_style_test_frameworks_js_1 = require("../helpers/mocha-style-test-frameworks.js");
const meta = __importStar(require("./generated-meta.js"));
const MOCHA_TEST_AND_SUITE_NAMES = [...mocha_style_test_frameworks_js_1.TEST_FUNCTION_NAMES, ...mocha_style_test_frameworks_js_1.SUITE_FUNCTION_NAMES];
const MESSAGE = 'Replace this empty or whitespace-only title with a descriptive name.';
const MESSAGE_ID = 'emptyTitle';
exports.rule = {
    meta: (0, generate_meta_js_1.generateMeta)(meta, {
        messages: {
            [MESSAGE_ID]: MESSAGE,
        },
    }),
    create(context) {
        if (!(0, module_js_1.importsOrDependsOnModule)(context, mocha_style_test_frameworks_js_1.SUPPORTED_TEST_FRAMEWORKS, mocha_style_test_frameworks_js_1.SUPPORTED_TEST_FRAMEWORKS)) {
            return {};
        }
        return {
            CallExpression(node) {
                if (isTestOrSuiteDeclaration(context, node)) {
                    checkTitle(context, node);
                }
            },
        };
    },
};
function checkTitle(context, node) {
    const titleNode = node.arguments[0];
    const title = titleNode && (0, mocha_style_test_frameworks_js_1.getStaticTitle)(titleNode);
    if (title?.trim() === '') {
        context.report({
            node: titleNode,
            messageId: MESSAGE_ID,
        });
    }
}
function isTestOrSuiteDeclaration(context, node) {
    if (!(0, mocha_style_test_frameworks_js_1.hasCallback)(node)) {
        return false;
    }
    return ((0, mocha_style_test_frameworks_js_1.isMochaTestConstruct)(context, node, MOCHA_TEST_AND_SUITE_NAMES) ||
        isPlaywrightTest(context, node) ||
        isPlaywrightDescribe(context, node));
}
function isPlaywrightTest(context, node) {
    const qualifiers = (0, mocha_style_test_frameworks_js_1.getPlaywrightTestQualifiers)(context, node.callee);
    return (qualifiers !== undefined &&
        qualifiers.every(qualifier => mocha_style_test_frameworks_js_1.PLAYWRIGHT_TEST_MODIFIERS.has(qualifier)));
}
function isPlaywrightDescribe(context, node) {
    const qualifiers = (0, mocha_style_test_frameworks_js_1.getPlaywrightTestQualifiers)(context, node.callee) ??
        (0, mocha_style_test_frameworks_js_1.getPlaywrightDescribeQualifiers)(node.callee);
    if (qualifiers?.[0] !== 'describe') {
        return false;
    }
    const modifiers = qualifiers.slice(1);
    const runnableModifiers = modifiers.at(-1) === mocha_style_test_frameworks_js_1.PLAYWRIGHT_DESCRIBE_FOCUS_MODIFIER ? modifiers.slice(0, -1) : modifiers;
    return runnableModifiers.every(modifier => mocha_style_test_frameworks_js_1.PLAYWRIGHT_DESCRIBE_MODIFIERS.has(modifier));
}
