// eslint-plugin-sonarjs/cjs/S5914/rule.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * You can redistribute and/or modify this program under the terms of
 * the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
// https://sonarsource.github.io/rspec/#/rspec/S5914/javascript
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const assertions_js_1 = require("../helpers/assertions.js");
const ast_js_1 = require("../helpers/ast.js");
const generate_meta_js_1 = require("../helpers/generate-meta.js");
const meta = __importStar(require("./generated-meta.js"));
const messages = {
    issue: 'Replace this assertion; it always succeeds or fails.',
    // hint targeted at identity comparisons against a freshly-created value: a deep equality matcher is almost always what was meant
    freshIdentity: 'Use a deep equality matcher; freshly-created values are never identical to other values.',
    // hint for predicate assertions on a freshly-created value, where the truthiness/nullishness is statically known
    freshPredicate: 'Replace this assertion; the value is freshly created here, so the result is independent of the code under test.',
};
exports.rule = {
    meta: (0, generate_meta_js_1.generateMeta)(meta, { messages }),
    create(context) {
        function checkAssertion(node) {
            const assertion = (0, assertions_js_1.extractTestAssertion)(context, node);
            const trivial = assertion ? getTrivialAssertion(context, assertion) : null;
            if (trivial) {
                context.report({ node: trivial.reportNode, messageId: trivial.messageId });
            }
        }
        return {
            CallExpression(node) {
                checkAssertion(node);
            },
            MemberExpression(node) {
                checkAssertion(node);
            },
        };
    },
};
/**
 * Classifies the assertion as trivial (always succeeds or always fails) and picks the most
 * informative message for the underlying cause.
 * @returns the node to report and the corresponding message id, or null if the assertion is not trivial.
 */
function getTrivialAssertion(context, assertion) {
    switch (assertion.kind) {
        case 'predicate':
            if (isConstantPrimitiveValue(context, assertion.actual)) {
                return { reportNode: assertion.actual, messageId: 'issue' };
            }
            if (isFreshReferenceExpression(assertion.actual)) {
                return { reportNode: assertion.actual, messageId: 'freshPredicate' };
            }
            return null;
        case 'comparison':
            if (assertion.comparison === 'strict') {
                // a freshly-created value on either side makes strict equality trivially fail
                // (or trivially succeed when negated). Loose equality is excluded because object-to-primitive
                // coercion can make comparisons like `[] == false` pass.
                if (isFreshReferenceExpression(assertion.actual)) {
                    return { reportNode: assertion.actual, messageId: 'freshIdentity' };
                }
                if (isFreshReferenceExpression(assertion.expected)) {
                    return { reportNode: assertion.expected, messageId: 'freshIdentity' };
                }
            }
            // both sides are constant (syntactically or via a resolved binding): for strict or loose
            // equality the comparison result is statically determined. Deep equality is intentionally
            // skipped via an explicit whitelist to avoid widening this rule when assertion helper
            // comparison kinds evolve.
            if ((assertion.comparison === 'strict' || assertion.comparison === 'loose') &&
                isConstantPrimitiveValue(context, assertion.actual) &&
                isConstantPrimitiveValue(context, assertion.expected)) {
                return { reportNode: assertion.actual, messageId: 'issue' };
            }
            return null;
        default:
            return null;
    }
}
const UNARY_OPERATORS_PRESERVING_CONSTNESS = new Set([
    '!',
    '+',
    '-',
    'typeof',
]);
function isConstantPrimitiveValue(context, node, visited) {
    // resolve `const` bindings through their initializer so that identifiers bound to a constant
    // primitive are treated as constants too. `visited` guards against mutually-recursive const
    // references like `const a = b; const b = a;` (TDZ at runtime).
    if (node.type === 'Identifier' && node.name !== 'undefined') {
        visited ??= new Set();
        const resolved = resolveConstBinding(context, node, visited);
        if (resolved) {
            return isConstantPrimitiveValue(context, resolved, visited);
        }
    }
    switch (node.type) {
        case 'Literal':
            return (node.value === null ||
                typeof node.value === 'boolean' ||
                typeof node.value === 'string' ||
                typeof node.value === 'number' ||
                ('bigint' in node && typeof node.bigint === 'string'));
        case 'Identifier':
            if (node.name === 'undefined') {
                const variable = (0, ast_js_1.getVariableFromName)(context, 'undefined', node);
                // global `undefined` has no user definitions; a shadowed binding does
                return !variable || variable.defs.length === 0;
            }
            return false;
        case 'TemplateLiteral':
            return node.expressions.length === 0;
        case 'UnaryExpression':
            // `void X` is always undefined regardless of X; the other operators preserve constness
            return (node.operator === 'void' ||
                (UNARY_OPERATORS_PRESERVING_CONSTNESS.has(node.operator) &&
                    isConstantPrimitiveValue(context, node.argument, visited)));
        case 'BinaryExpression':
        case 'LogicalExpression':
            return (isConstantPrimitiveValue(context, node.left, visited) &&
                isConstantPrimitiveValue(context, node.right, visited));
        default:
            return false;
    }
}
/**
 * Returns the initializer of `node` if it refers to a `const` binding declared with a plain identifier
 * pattern (no destructuring) and that binding has not been visited already, otherwise undefined.
 * Restricting to `const` ensures the binding's value is statically equal to the initializer at every read.
 * The variable is added to `visited` so mutually-recursive references cannot loop forever.
 */
function resolveConstBinding(context, node, visited) {
    const variable = (0, ast_js_1.getVariableFromName)(context, node.name, node);
    if (!variable || visited.has(variable) || variable.defs.length !== 1) {
        return undefined;
    }
    const def = variable.defs[0];
    if (def.type !== 'Variable' || def.parent?.type !== 'VariableDeclaration') {
        return undefined;
    }
    if (def.parent.kind !== 'const' || def.node.id.type !== 'Identifier') {
        return undefined;
    }
    visited.add(variable);
    return def.node.init;
}
/**
 * These expression types create a new reference on each evaluation, so they are always different from any other value
 */
const FRESH_REFERENCE_EXPRESSIONS = new Set([
    'ArrayExpression',
    'ObjectExpression',
    'FunctionExpression',
    'ArrowFunctionExpression',
    'ClassExpression',
    'NewExpression',
]);
/**
 * Checks if the given node is an expression that creates a new reference on each evaluation, and thus is always different from any other value.
 */
function isFreshReferenceExpression(node) {
    // regex literals create a fresh RegExp object on each evaluation
    if (node.type === 'Literal' && 'regex' in node) {
        return true;
    }
    return FRESH_REFERENCE_EXPRESSIONS.has(node.type);
}
