// eslint-plugin-sonarjs/cjs/S3798/rule.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = void 0;
const generate_meta_js_1 = require("../helpers/generate-meta.js");
const ancestor_js_1 = require("../helpers/ancestor.js");
const meta = __importStar(require("./generated-meta.js"));
exports.rule = {
    meta: (0, generate_meta_js_1.generateMeta)(meta, {
        messages: {
            defineLocally: 'Define this declaration in a local scope or bind explicitly the property to the global object.',
        },
    }),
    create(context) {
        return {
            Program(node) {
                if (isModuleFile(context, node)) {
                    return;
                }
                const globalScope = context.sourceCode.scopeManager.globalScope;
                const topLevelScope = findTopLevelScope(context);
                if (!globalScope || !topLevelScope) {
                    return;
                }
                const predefinedGlobals = getPredefinedGlobals(context, globalScope);
                for (const variable of topLevelScope.variables) {
                    processVariable(context, variable, predefinedGlobals);
                }
            },
        };
    },
};
function processVariable(context, variable, predefinedGlobals) {
    if (predefinedGlobals.has(variable.name)) {
        // Avoid reporting on redefinitions of known globals such as isNaN or alert.
        return;
    }
    for (const def of variable.defs) {
        const defNode = def.node;
        if (def.type === 'FunctionName' || (def.type === 'Variable' && def.parent?.kind === 'var')) {
            context.report({
                node: defNode,
                messageId: 'defineLocally',
            });
            return;
        }
    }
}
function isModuleFile(context, node) {
    const detectedModuleType = getDetectedModuleType(context);
    return (detectedModuleType === 'module' ||
        detectedModuleType === 'commonjs' ||
        usesModuleSyntax(node, context.sourceCode));
}
function findTopLevelScope(context) {
    return (context.sourceCode.scopeManager.scopes.find(scope => scope.type === 'module') ??
        context.sourceCode.scopeManager.globalScope);
}
function getPredefinedGlobals(context, globalScope) {
    const settings = context.settings;
    return new Set([
        ...(settings.predefinedGlobals ?? []),
        ...globalScope.variables.filter(variable => variable.defs.length === 0).map(({ name }) => name),
    ]);
}
function getDetectedModuleType(context) {
    return context.settings.detectedModuleType;
}
function usesModuleSyntax(node, sourceCode) {
    return node.body.some(statement => isModuleDeclaration(statement) || hasTopLevelAwait(statement, sourceCode));
}
function isModuleDeclaration(statement) {
    return (statement.type === 'ImportDeclaration' ||
        statement.type === 'ExportAllDeclaration' ||
        statement.type === 'ExportDefaultDeclaration' ||
        statement.type === 'ExportNamedDeclaration');
}
function hasTopLevelAwait(node, sourceCode) {
    if (node.type === 'ArrowFunctionExpression' ||
        node.type === 'FunctionDeclaration' ||
        node.type === 'FunctionExpression') {
        return false;
    }
    if (node.type === 'AwaitExpression') {
        return true;
    }
    if (node.type === 'ForOfStatement' && node.await) {
        return true;
    }
    return (0, ancestor_js_1.childrenOf)(node, sourceCode.visitorKeys).some(child => hasTopLevelAwait(child, sourceCode));
}
