// eslint-plugin-sonarjs/cjs/helpers/regex/extract.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getParsedRegex = getParsedRegex;
exports.getPatternFromNode = getPatternFromNode;
const regexpp = __importStar(require("@eslint-community/regexpp"));
const ast_js_1 = require("../ast.js");
const ast_js_2 = require("./ast.js");
const flags_js_1 = require("./flags.js");
function getParsedRegex(node, context) {
    const patternAndFlags = getPatternFromNode(node, context);
    if (patternAndFlags) {
        try {
            return regexpp.parseRegExpLiteral(new RegExp(patternAndFlags.pattern, patternAndFlags.flags));
        }
        catch {
            // do nothing for invalid regex
        }
    }
    return null;
}
function getPatternFromNode(node, context) {
    if ((0, ast_js_2.isRegExpConstructor)(node)) {
        const patternOnly = getPatternFromNode(node.arguments[0], context);
        const flags = (0, flags_js_1.getFlags)(node, context);
        if (patternOnly && flags !== null) {
            // if we can't extract flags correctly, we skip this
            // to avoid FPs
            return { ...patternOnly, flags };
        }
    }
    else if ((0, ast_js_1.isRegexLiteral)(node)) {
        return literalPattern(node.regex.pattern, node.regex.flags);
    }
    else if ((0, ast_js_1.isStringLiteral)(node)) {
        return literalPattern(node.value);
    }
    else if ((0, ast_js_1.isStaticTemplateLiteral)(node)) {
        return literalPattern(node.quasis[0].value.raw);
    }
    else if ((0, ast_js_1.isSimpleRawString)(node)) {
        return literalPattern((0, ast_js_1.getSimpleRawStringValue)(node));
    }
    else if ((0, ast_js_1.isIdentifier)(node)) {
        const assignedExpression = (0, ast_js_1.getUniqueWriteUsage)(context, node.name, node);
        if (assignedExpression &&
            assignedExpression.parent?.type === 'VariableDeclarator') {
            const pattern = getPatternFromNode(assignedExpression, context);
            return pattern ? { ...pattern, isPureLiteral: false } : null;
        }
    }
    else if ((0, ast_js_1.isBinaryPlus)(node)) {
        const left = getPatternFromNode(node.left, context);
        const right = getPatternFromNode(node.right, context);
        return left && right ? concatPatterns(left, right) : null;
    }
    return null;
}
function literalPattern(pattern, flags = '') {
    return { pattern, flags, seams: [], isPureLiteral: true };
}
function concatPatterns(left, right) {
    const splitOffset = left.pattern.length;
    const splitSeams = left.isPureLiteral && right.isPureLiteral ? [] : [splitOffset];
    return {
        pattern: left.pattern + right.pattern,
        flags: '',
        seams: [...left.seams, ...splitSeams, ...right.seams.map(s => s + splitOffset)],
        isPureLiteral: left.isPureLiteral && right.isPureLiteral,
    };
}
